### ISSUE TITLE
How to add dynamically allocated strings to Pickler?

### ISSUE THREAD

**Issue by mrshenli** [pytorch/pytorch#20090]
The following code prints `111` and `111`, instead of `222` and `111`, because `222` is skipped [here](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/pickler.cpp#L68). Is this by design as Pickler only works for statically allocated strings? Or is there a way to correctly add dynamically allocated strings? (and all other types listed [here](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/pickler.cpp#L104-L114))? 

```c++
  std::string str1 = "111";
  std::string str2 = "222";

  std::vector<at::Tensor> tensor_table;
  torch::jit::Pickler pickler(&tensor_table);
  pickler.start();
  pickler.addIValue(str1);
  pickler.addIValue(str2);
  pickler.finish();

  auto buffer = new char[pickler.stack().size()];
  memcpy(buffer, pickler.stack().data(), pickler.stack().size());

  torch::jit::Unpickler unpickler(buffer, pickler.stack().size(), &tensor_table);
  auto values = unpickler.parse_ivalue_list();
  std::cout << values.back().toStringRef() << std::endl;
  values.pop_back();
  std::cout << values.back().toStringRef() << std::endl;
  values.pop_back();
```

cc @zdevito

**Comment by suo** [comment]
cc @driazati

**Comment by driazati** [comment]
This should be fixed with #18154, feel free to re-open if it's still an issue
